apps/dav/lib/Connector/Sabre/ZipFolderPlugin.php

```diff
@@ -0,0 +1,136 @@
+ */
+namespace OCA\DAV\Connector\Sabre;
+
+use OC\Streamer;
+use OCP\Files\File as NcFile;
+use OCP\Files\Folder as NcFolder;
+use OCP\Files\Node as NcNode;
+use Psr\Log\LoggerInterface;
+use Sabre\DAV\ICollection;
+use Sabre\DAV\ServerPlugin;
+use Sabre\DAV\Tree;
+use Sabre\HTTP\Request;
+use Sabre\HTTP\Response;
+
+/**
+ * This plugin allows to download folders accessed by GET HTTP requests on DAV.
+ * The WebDAV standard explicitly say that GET is not covered and should return what ever the application thinks would be a good representation.
+ *
+ * When a collection is accessed using GET, this will provide the content as a archive.
+ * The type can be set by the `Accept` header (MIME type of zip or tar), or as browser fallback using a `accept` GET parameter.
+ * It is also possible to only include some child nodes (from the collection it self) by providing a `filter` GET parameter or `X-NC-Files` custom header.
+ */
+class ZipFolderPlugin extends ServerPlugin {
+
+	/**
+	 * Reference to main server object
+	 *
+	 * @var Server
+	 */
+	private $server;
+
+	public function __construct(
+		private Tree $tree,
+		private LoggerInterface $logger,
+	) {
+	}
+
+	/**
+	 * This initializes the plugin.
+	 *
+	 * This function is called by \Sabre\DAV\Server, after
+	 * addPlugin is called.
+	 *
+	 * This method should set up the required event subscriptions.
+	 *
+	 * @param \Sabre\DAV\Server $server
+	 * @return void
+	 */
+	public function initialize(\Sabre\DAV\Server $server) {
+		$this->server = $server;
+		$this->server->on('method:GET', [$this, 'handleDownload'], 100);
+	}
+
+	/**
+	 * Adding a node to the archive streamer.
+	 * This will recursively add new nodes to the stream if the node is a directory.
+	 */
+	protected function streamNode(Streamer $streamer, NcNode $node, string $rootPath): void {
+		// Remove the root path from the filename to make it relative to the requested folder
+		$filename = str_replace($rootPath, '', $node->getPath());
+
+		if ($node instanceof NcFile) {
+			$resource = $node->fopen('rb');
+			if ($resource === false) {
+				$this->logger->info('Cannot read file for zip stream', ['filePath' => $node->getPath()]);
+				throw new \Sabre\DAV\Exception\ServiceUnavailable('Requested file can currently not be accessed.');
+			}
+			$streamer->addFileFromStream($resource, $filename, $node->getSize(), $node->getMTime());
+		} elseif ($node instanceof NcFolder) {
+			$streamer->addEmptyDir($filename);
+			$content = $node->getDirectoryListing();
+			foreach ($content as $subNode) {
+				$this->streamNode($streamer, $subNode, $rootPath);
+			}
+		}
+	}
+
+	public function handleDownload(Request $request, Response $response): null|false {
+		$node = $this->tree->getNodeForPath($request->getPath());
+		if (!($node instanceof ICollection)) {
+			// skip if not a collection
+			return null;
+		}
+
+		if (!($node instanceof \OCA\DAV\Connector\Sabre\Directory)) {
+			// only handle directories
+			return null;
+		}
```

Comment:
```suggestion
		if (!($node instanceof \OCA\DAV\Connector\Sabre\Directory)) {
			// only handle directories
			return null;
		}
```
directory is always a collection?